### PR TITLE
[#5105] improvement(server,client): Error code optimization about access control API

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?

Error code optimization about access control API

### Why are the changes needed?

Fix: #5105 

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?

Modify some UTs

### CODE REVIEW

File: api/src/main/java/org/apache/gravitino/exceptions/IllegalMetadataObjectException.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.exceptions;
+
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+/** An exception thrown when a metadata object is invalid. */
+public class IllegalMetadataObjectException extends IllegalArgumentException {
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/exceptions/IllegalMetadataObjectException.java:25]: How do you define "illegal"?

File: server/src/main/java/org/apache/gravitino/server/web/rest/PermissionOperations.java [hunk 2]
```diff
@@ -87,12 +89,17 @@ public Response grantRolesToUser(
...
-                              Utils.ok(
+                          () -> {
+                            try {
+                              return Utils.ok(
                                   new UserResponse(
                                       DTOConverters.toDTO(
                                           accessControlManager.grantRolesToUser(
-                                              metalake, request.getRoleNames(), user)))))));
+                                              metalake, request.getRoleNames(), user))));
+                            } catch (NoSuchRoleException nsr) {
```
Comment (jerryshao) [inline server/src/main/java/org/apache/gravitino/server/web/rest/PermissionOperations.java:99]: I think it is better to handle `IllegalRoleException` in the core, not convert to this exception in here.

File: api/src/main/java/org/apache/gravitino/exceptions/IllegalMetadataObjectException.java (hunk 1 above)
Comment (jerqi) [reply api/src/main/java/org/apache/gravitino/exceptions/IllegalMetadataObjectException.java:25]: If a metadata object in the path doesn't exist, we will return NotFoundException.
If a metadata object in the request body doesn't exist, we will return IllegalException.

File: server/src/main/java/org/apache/gravitino/server/web/rest/PermissionOperations.java (hunk 2 above)
Comment (jerqi) [reply server/src/main/java/org/apache/gravitino/server/web/rest/PermissionOperations.java:99]: OK , just in my view, I usually throw `IllegalxxxException` in the other `Operation` class.  It's ok to throw in the core module.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**